pkg/system/autoscaler.go

```diff
@@ -0,0 +1,239 @@
+package system
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	nbv1 "github.com/noobaa/noobaa-operator/v5/pkg/apis/noobaa/v1alpha1"
+	"github.com/noobaa/noobaa-operator/v5/pkg/bundle"
+	"github.com/noobaa/noobaa-operator/v5/pkg/nb"
+	"github.com/noobaa/noobaa-operator/v5/pkg/options"
+	"github.com/noobaa/noobaa-operator/v5/pkg/util"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/sirupsen/logrus"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	kedaHpaNoobaaEndpoint = "keda-hpa-noobaa-endpoint"
+	noobaaEndpoint        = "noobaa-endpoint"
+)
+
+func (r *Reconciler) reconsileAutoscaler() error {
+	log := r.Logger.WithField("func", "ReconcileHPAEndpoint")
+	var err error
+	prometheusList := deployedPrometheus(log)
+	autoscalingType := options.DefaultAutoscalingType
+	if r.NooBaa.Spec.AutoscalingType != "" {
+		autoscalingType = string(r.NooBaa.Spec.AutoscalingType)
+	}
+	if err = r.cleanPreviousAutoscalingResources(string(r.NooBaa.Spec.AutoscalingType)); err != nil {
+		return err
+	}
+	switch autoscalingType {
+	case string(nbv1.AutoscalingTypeKeda):
+		if err = r.autoscaleKeda(prometheusList); err != nil {
+			return err
+		}
+	case string(nbv1.AutoscalingTypeHPAV2):
+		//TODO
+		log.Infof("HPAV2 implementation pending")
+	case string(nbv1.AutoscalingTypeHPAV1):
+		if err = r.ReconcileObject(r.HPAEndpoint, r.SetDesiredHPAEndpoint); err != nil {
+			return err
+		}
+	default:
+		log.Infof("❌ Autoscaling type is not defined for noobaa")
+		return errors.New("Autoscaling type is not defined")
+	}
+	return nil
+}
+
+func (r *Reconciler) cleanPreviousAutoscalingResources(autoscalingType string) error {
+	autoscalers := &autoscalingv1.HorizontalPodAutoscalerList{}
+	util.KubeList(autoscalers, &client.ListOptions{Namespace: r.Request.Namespace})
+	if len(autoscalers.Items) == 0 {
+		return nil
+	}
```

Comment:
@dannyzaken we need to delete the existing HPA even for Keda and HPAV2 for that we need to get the existing autoscaler, correct me if am wrong

Reply:
@dannyzaken I could see that util.KubeList() is listing all HPAs even with version `autoscaling/v2`. This code can delete all the HPA with  `apiVersion` such as `autoscaling/v2`, `autoscaling/v2beta2` and `autoscaling/v1`.

Reply:
@naveenpaul1 That's weird since in `KubeList` we list for specific `GroupVersionKind` and also `autoscalers` is of type `autoscalingv1.HorizontalPodAutoscalerList`. can you validate? if this is indeed the behavior then it seems like a bug to me

Reply:
@dannyzaken  autoscalingv1 and autoscalingv2 check is added to the code before deleting HPAs, PR updated